### ISSUE TITLE
Fine tune for smaller objects

### ISSUE THREAD

**Issue by groot-1313** [pjreddie/darknet#403]
I have a dataset with bigger as well as smaller objects. Upon testing with a pre trained model on my own dataset, YOLO performs reasonably well(some detected correctly, but wrong classes), but I would like to improve the accuracy by fine-tuning the model to my own dataset. I also have images with smaller objects that have gone undetected. I think this has to do with the anchor size. 

My current approach is to freeze all but the last few layers, and fine tune it on my own data. And to make sure the smaller objects are also detected, I think adding a few more anchors of smaller size should work. 
I was wondering if anymore changes are required? Any inputs will be very much appreciated!

**Comment by Nopileos2** [comment]
The anchors should be generated from your dataset with e.g. kmeans. You can try this script. 
https://github.com/Jumabek/darknet_scripts/blob/master/gen_anchors.py

It is important that the anchors are generated from your dataset, since yolo only predicts offsets. I would try it with 5 anchors and increase it if you feel like 5 are not enough for your dataset, because you have a lot of different sized objects.

**Comment by groot-1313** [comment]
Thanks! I will use the script. 

How do I go about freezing the layers? 

And would that be a good idea? Because what I had in mind was to keep the layers that detect important low level features the fixed in terms of weight and only train the last few layers that detect high level features. Or would it be advisable to train all the layers using a pre trained model?

**Comment by Nopileos2** [comment]
I don't know how your dataset looks like but in general the first few layers often learn the same features as long as they are trained on comparable data. I would first try training without any freezing, it maybe takes a bit longer but should work.

Freezing would need some coding. You have to rewrite some code which would disable the weight updates for specific layers. Don't know how familiar you are with the code so this task could take a while.

**Comment by groot-1313** [comment]
oh alright. I will try without freezing first. 

[https://gist.github.com/StevenPuttemans/fb93a7962cc0157d4c1a60014b1775a9](url) this link mentions how to freeze the top network by mentioning the last layer to be freezed. Though one might have to go through the code to know where to make the changes. I am unsure of whether any changes to the cfg file would enable freezing of the layers.

**Comment by groot-1313** [comment]
@Nopileos2 Yes, but it should be set in the source files right? I haven't gone through the code. Do you know which file this parameter is to be set in?

**Comment by Nopileos2** [comment]
In parser.c the option is read https://github.com/pjreddie/darknet/blob/80d9bec20f0a44ab07616215c6eadb2d633492fe/src/parser.c#L724

so it is a parameter for every layer if i read it correctly so you write it into the .cfg "in" a layer just like "batch_normalize=1". So you would write "stopbackward=1" and it would stop the backward process at this layer, so all layers before this layer are also not updated. This you can see in network.c 
https://github.com/pjreddie/darknet/blob/80d9bec20f0a44ab07616215c6eadb2d633492fe/src/network.c#L272

**Comment by groot-1313** [comment]
Oh, that helps!! Thanks a lot @Nopileos2

**Comment by b-miethig** [comment]
Hi,

I tried using stopbackward with yoloV2 but it doesn't seem to be working. Did this work for you? If so, what exactly did you do?

[net]
batch=64
subdivisions=8
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
max_batches = 40100
policy=steps
steps=-1,100,20000,30000
scales=.1,10,.1,.1

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

###########

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky
# Added in stopbackward to avoid training layers above this
stopbackward=1

[convolutional]
size=1
stride=1
pad=1
filters=50
activation=linear

[region]
anchors = 1.08,1.19,  3.42,4.41,  6.63,11.38,  9.42,5.11,  16.62,10.52
bias_match=1
classes=5
coords=4
num=5
softmax=1
jitter=.2
rescore=1

object_scale=5
noobject_scale=1
class_scale=1
coord_scale=1

absolute=1
thresh = .5
random=1

**Comment by luochonghai** [comment]
Many thanks for your great work!

**Comment by ibrahimLearning** [comment]
@Nopileos2 
I am modified according to your advice, I added this line in parser.c:
`l.stopbackward = option_find_int_quiet(options, "stopbackward", 0);` 
After modification, when I want to build it using "make" , it gives the following error:
"./src/network.c:221:13: error: 'layer {aka struct layer}' has no member named 'stopbackward'
if(l.stopbackward) break; "
^
Would you give some advice please?

**Comment by Nopileos2** [comment]
There should be no need to add anything. the line should already exists in parser.c and the option should already be defined in the layer struct in darknet.h. All you have to do is add it into your config at the desired layer, where the backward process should stop.

If not please see which version you are working on and see if you can update to the latest one. 

For reference:
https://github.com/pjreddie/darknet/blob/61c9d02ec461e30d55762ec7669d6a1d3c356fb2/include/darknet.h#L200

https://github.com/pjreddie/darknet/blob/61c9d02ec461e30d55762ec7669d6a1d3c356fb2/src/parser.c#L843

**Comment by ibrahimLearning** [comment]
@Nopileos2 Thanks I catch your point. Actually I am using different and modified algorithm from yolo. But I catch your point.
Thanks

**Comment by kxlwolf** [comment]
> There should be no need to add anything. the line should already exists in parser.c and the option should already be defined in the layer struct in darknet.h. All you have to do is add it into your config at the desired layer, where the backward process should stop.
> 
> If not please see which version you are working on and see if you can update to the latest one.
> 
> For reference:
> https://github.com/pjreddie/darknet/blob/61c9d02ec461e30d55762ec7669d6a1d3c356fb2/include/darknet.h#L200
> 
> https://github.com/pjreddie/darknet/blob/61c9d02ec461e30d55762ec7669d6a1d3c356fb2/src/parser.c#L843

thanks a lot!

**Comment by dungdo123** [comment]
> @Nopileos2 Thanks I catch your point. Actually I am using different and modified algorithm from yolo. But I catch your point.
> Thanks

HI @Nopileos2, can you talk more about your modified algorithm?
Now I am trying to train the yolov4 model to detect person, helmet, and mask objects! My trained model performs well with normal size objects, but it can not detect well on small people as the author's trained model. I followed this tutorial (https://github.com/AlexeyAB/ScaledYOLOv4)
